Rewrote mailer getting start guide and reference documentation (#18540)
quarkusio/quarkus: docs/src/main/asciidoc/mailer-reference.adoc

* `io.quarkus.mailer.reactive.ReactiveMailer` provides the reactive (non-blocking and asynchronous) API

NOTE: The two APIs are equivalent feature-wise. Actually the `Mailer` implementation is built on top of the `ReactiveMailer` implementation.

[NOTE]
.Deprecation
====
`io.quarkus.mailer.ReactiveMailer` is deprecated in favor of `io.quarkus.mailer.reactive.ReactiveMailer`.
====

To send a simple email, proceed as follows:

[source, java]
----
// Imperative API:
mailer.send(Mail.withText("to@acme.org", "A simple email from quarkus", "This is my body."));
// Reactive API:
Uni<Void> stage = reactiveMailer.send(Mail.withText("to@acme.org", "A reactive email from quarkus", "This is my body."));
----

For example, you can use the `Mailer` in an HTTP endpoint as follows:

[source, java]
----
@GET
@Path("/imperative")
public void sendASimpleEmail() {
    mailer.send(Mail.withText("to@acme.org", "A simple email from quarkus", "This is my body"));
}

@GET
@Path("/reactive")
public Uni<Void> sendASimpleEmailAsync() {
    return reactiveMailer.send(
            Mail.withText("to@acme.org", "A reactive email from quarkus", "This is my body"));
}
----

== Creating Mail objects

> The mailer lets you send `io.quarkus.mailer.Mail` object.

Reviewer: ```suggestion
The mailer lets you send `io.quarkus.mailer.Mail` objects.
```
?

You can also send several `Mail` objects in one call:

[source, java]
----
mailer.send(mail1, mail2, mail3);
----

[[attachments]]
== Sending attachments

To send attachments, just use the `addAttachment` methods on the `io.quarkus.mailer.Mail` instance:

[source,java]
----
@GET
@Path("/attachment")
public void sendEmailWithAttachment() {
        mailer.send(Mail.withText("clement.escoffier@gmail.com", "An email from quarkus with attachment",
                "This is my body")
                .addAttachment("my-file-1.txt",
                        "content of my file".getBytes(), "text/plain")
                .addAttachment("my-file-2.txt",
                        new File("my-file.txt"), "text/plain")
        );
}
----

Attachments can be created from raw bytes (as shown in the snippet) or files.
Note that files are resolved from the working directory of the application.

[[html]]
== Sending HTML emails with inlined attachments

When sending HTML emails, you can add inlined attachments.
For example, you can send an image with your email, and this image will be displayed in the mail content.

> If you put the image file into resources folder, you should specify the full path to the file, _e.g._ `META-INF/resources/quarkus-logo.png` otherwise Quarkus will look for the file in the root directory.

Reviewer: ```suggestion
If you put the image file into the `META-INF/resources` folder, you should specify the full path to the file, _e.g._ `META-INF/resources/quarkus-logo.png` otherwise Quarkus will look for the file in the root directory.
```

TIP: Injected mail templates are validated during build.
The build fails if there is no matching template in `src/main/resources/templates`.

You can also do this without type-safe templates:

[source, java]
----
@Inject
@Location("hello")
MailTemplate hello; // <1>

@GET
@Path("/mail")
public Uni<Void> send() {
    return hello.to("to@acme.org") // <2>
       .subject("Hello from Qute template")
       .data("name", "John") // <3>
       .send() // <4>
}
----
<1> If there is no `@Location` qualifier provided, the field name is used to locate the template.
Otherwise, search for the template as the specified location. In this particular case, we will use the `src/main/resources/templates/hello.html` and `src/main/resources/templates/hello.txt` templates to create the message body.
<2> Create a mail template instance and set the recipient.
<3> Set the data used in the template.
<4> `MailTemplate.send()` triggers the rendering and, once finished, sends the e-mail via a `Mailer` instance.

[[testing]]
== Testing email sending

Because it is very inconvenient to send emails during development and testing, you can set the `quarkus.mailer.mock` boolean

> configuration to `true` to not actually send emails but print them on stdout and collect them in a `MockMailbox` bean instead.

Reviewer 2: ```suggestion
configuration to `true` to prevent the actual sending of emails but instead print them on stdout and collect them in a `MockMailbox` bean instead.
```
